packages/flutter/lib/src/rendering/object.dart

```diff
@@ -2605,6 +2596,41 @@ abstract class RenderObject extends AbstractNode with DiagnosticableTreeMixin im
       );
     }
   }
+
+
+  /// Clip [Canvas] with [T] ([Path], [RRect], or [Rect]) according to [Clip]
+  /// and then paint. [Canvas] is restored to the pre-clip status afterwards.
+  static void clipAndPaint<T>(Canvas canvas, Clip clip, T t, Rect bounds, void painter()) {
+    void Function(bool doAA) canvasClipCall;
+    if (t is Path) {
+      canvasClipCall = (bool doAA) => canvas.clipPath(t, doAntiAlias: doAA);
+    } else if (t is RRect) {
+      canvasClipCall = (bool doAA) => canvas.clipRRect(t, doAntiAlias: doAA);
+    } else if (t is Rect) {
+      canvasClipCall = (bool doAA) => canvas.clipRect(t, doAntiAlias: doAA);
+    }
+    assert(canvasClipCall != null); // t should be one of Path, RRect, Rect
+    canvas.save();
+    switch (clip) {
+      case Clip.none:
+        break;
+      case Clip.hardEdge:
+        canvasClipCall(false);
+        break;
+      case Clip.antiAlias:
+        canvasClipCall(true);
+        break;
+      case Clip.antiAliasWithSaveLayer:
+        canvasClipCall(true);
+        canvas.saveLayer(bounds, new Paint());
+        break;
+    }
+    painter();
+    if (clip == Clip.antiAliasWithSaveLayer) {
+      canvas.restore();
+    }
+    canvas.restore();
+  }
```

Comment:
This belongs in the painting layer.

I recommend splitting it out into three versions so you get better type checking (they can still all be implemented in terms of a private version that uses the closure like you do here).

Reply:
Done.